File: code du prof/Idefix-2024/Idefix/python/tools/idefix_server.py
import cv2
import numpy as np
import argparse
import os
import sys

absolute_path = os.path.dirname(__file__)
relative_path = "../modules"
full_path = os.path.join(absolute_path, relative_path)
sys.path.append(full_path)

from networking import Buffer, TCPServerAbstraction
from encoding import Packer

from jpeg_traits import JpegImage

class Server(TCPServerAbstraction):
    def __init__(self):
        super().__init__(2048)
    def onConnected(self,client):
        print('connexion d\'un client : ',client[1])
        self.sendTo(client,sizeBuffer)
        # all clients require broadcast
        return True
    def onDisconnected(self,client):
        print('deconnexion d\'un client : ',client[1])
    def start(self,args):
        self.initialize(args.interface,args.port)
        self.passiveReceive()
        self.listenToClients()
    def stop(self):
        self.finalize()

captureDeviceIds = [0, 1, 2, 3]

WIDTH   = 1280
HEIGHT  = 720
SCALE   = 1

size = [HEIGHT, WIDTH, 3]
frame = np.zeros(size,np.uint8)
millisecondsToWait = int(1000 // 30)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('-i', '--interface', action='store', default='', type=str, help='network interface to use')
    parser.add_argument('-p', '--port', action='store', default=2120, type=int, help='port on server')
    args = parser.parse_args()
    
    sizeBuffer = Buffer(Packer.pack(size))
    frameBuffer = Buffer(Packer.pack(frame,'cv::Mat'))
    endLoop = False
    
    captureDevices = []
    for id in captureDeviceIds:
        cap = cv2.VideoCapture(id)
        if not cap.isOpened():
            endLoop = True
            break
        cap.set(cv2.CAP_PROP_FRAME_WIDTH, WIDTH)
        cap.set(cv2.CAP_PROP_FRAME_HEIGHT, HEIGHT)
        print("Cap {:0d} opened {:0}x{:0}".format(id,cap.get(cv2.CAP_PROP_FRAME_WIDTH),cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
        captureDevices.append(cap)

    server = Server ()
    server.start(args)
    cv2.namedWindow('frame',cv2.WINDOW_FREERATIO)
    while not endLoop:
        capturedFrames = []
        for cap in captureDevices:
            ret, frame = cap.read()
            if not ret:
                print("Can't receive frame (stream end?). Exiting ...")
                endLoop = True
                break
            capturedFrames.append(cv2.resize(frame,(WIDTH>>SCALE,HEIGHT>>SCALE)))
        if endLoop: break
        while len(capturedFrames) < 4:
            capturedFrames.append(np.zeros_like(capturedFrames[0]))
        row1 = None
        row2 = None
        if len(captureDeviceIds)==2:
            row1 = capturedFrames[0]
            row2 = capturedFrames[1]
        else:
            row1 = np.hstack((capturedFrames[0],capturedFrames[1]))
            row2 = np.hstack((capturedFrames[2],capturedFrames[3]))
        frame = np.vstack((row1,row2))
        cv2.imshow('frame', frame)
        if server.nClients>0:
            frameBuffer = Buffer(Packer.pack(frame,'cv::Mat'))
            server.broadcast(frameBuffer)
        key = cv2.waitKey(millisecondsToWait) & 0x0FF
        if key == ord('q'): break

    for cap in captureDevices:
        cap.release()

    cv2.destroyAllWindows()

File: code du prof/Idefix-2024/Idefix/c++/tools/idefix_server/srcs/idefix_server.cpp
#include <iostream>
#include <iterator>
#include <map>

#include <opencv2/opencv.hpp>

#include "networking/TCPServer.h"

#include "ArgumentParser.h"
#include "PackImage.h"

const unsigned int WIDTH    = 1920;
const unsigned int HEIGHT   = 1080;
const unsigned int FRATE    = 30;

class Server : private networking::TCPServerAbstraction
{
public:
    Server() : networking::TCPServerAbstraction(2048) {
        size.push_back(HEIGHT);
        size.push_back(WIDTH);
        size.push_back(3);
        encoding::Bytearray ba = encoding::Packer::pack(size);
        sizeBuffer = networking::Buffer(ba);
    }
    bool clientIsConnecting(networking::ClientData datas) {
        auto [sock, info, cnx] = datas;
        sendTo(sock,sizeBuffer);
        return true;
    }
    void messageReceived(networking::ClientData datas,networking::Buffer buffer) {
        auto [sock, info, cnx] = datas;
        encoding::Bytearray ba(buffer.data(),buffer.data()+buffer.length());
        auto [index, result] = encoding::Packer::unpack(ba,0);
        unsigned int value = result.value<unsigned int>();
        sendTo(sock,dataBuffer);
    }
    void start(ParsedArgs args)
    {
        std::string interface(std::any_cast<std::string>(args["interface"]));
        int port = std::any_cast<int>(args["port"]);
        initialize(interface,port);
        passiveReceive([this](networking::ClientData d,networking::Buffer b){this->messageReceived(d,b);});
        listenToClients([this](networking::ClientData d)->bool{return this->clientIsConnecting(d);});
    }
    void stop()
    {
        finalize();
    }
    void sendAll(networking::Buffer b) {
    }
    void setFrame(const cv::Mat& frame) {
        dataBuffer = networking::Buffer(encoding::Packer::pack(frame))
    }
    encoding::List size;
    networking::Buffer sizeBuffer;
    networking::Buffer dataBuffer;
};

cv::VideoCapture openWebcam(unsigned int id, unsigned int width, unsigned int height) {
    cv::VideoCapture cap(id);
    cap.set(CAP_PROP_FRAME_HEIGHT, height);
    cap.set(CAP_PROP_FRAME_WIDTH, width);
    return cap;
}

int main(int argc, char** argv)
{
    ArgumentParser parser;
    parser.add_argument<std::string>('i',"interface",std::string(""),"store",true,"");
    parser.add_argument<int>('p',"port",12345,"store",true,"");
    ParsedArgs args = parser.parse_args(argc,argv);
    Server server;
    std::cout << "pre start" << std::endl;
    server.start(args);
    std::cout << "port start" << std::endl;
    cv::namedWindow("emitted frame");
    cv::VideoCapture cap0 = openWebcam(0, WIDTH, HEIGHT);
    cv::VideoCapture cap1 = openWebcam(1, WIDTH, HEIGHT);
    cv::VideoCapture cap2 = openWebcam(2, WIDTH, HEIGHT);
    cv::Mat frame0, frame1, frame2, frame;
    cap0.read(frame0); cap1.read(frame1), cap2.read(frame2);
    frame = mergeFrames(frame0, frame1, frame2);
    server._frame = frame;
    while(true)
    {
        cap0.read(frame0); cap1.read(frame1), cap2.read(frame2);
        frame = mergeFrames(frame0, frame1, frame2); 
        server.setFrame(frame);
        cv::imshow("emitted frame",frame);
        int key = cv::waitKey(FRATE) & 0x0FF;
        if (key == 'x') break;
    }
    server.stop();
    return 0;
}